### ISSUE TITLE
Spellchecking documentation in CI

### ISSUE THREAD

**Issue by tbroadley** [eslint/eslint#9793]
I've created [a command-line tool](https://github.com/tbroadley/spellchecker-cli) that checks Markdown and text files for spelling mistakes and some types of grammatical errors. I think it could be useful to run this tool on ESLint's documentation and other Markdown files (_e.g._ `README.md`) during CI. If you agree, I'm happy to open a PR.

**Comment by platinumazure** [comment]
Hi @tbroadley, thanks for the issue! I'm interested to see how this would look.

PRs that fix misspellings in existing docs should be accepted with no or very little controversy.

I don't think we would want to add spellchecker to our build process without our Technical Steering Committee discussing it. As I said before, though, I'm interested in checking it out (and I might do so locally in the near future).

**Comment by tbroadley** [comment]
Alright, thanks for the reply!

**Comment by j-f1** [comment]
FYI I used the tool for #9751, and @tbroadley used it for #9723.

**Comment by tbroadley** [comment]
@j-f1 This is actually a different tool from [that one](https://github.com/tbroadley/github-spellcheck-cli), but the idea is pretty much the same! This tool is set up to just list spelling mistakes, rather than allowing the user to fix them and create a PR.

**Comment by platinumazure** [comment]
@tbroadley I've just played around with this locally and I have a few questions:

1. Is there a way to specify (either in the dictionary or in some other configuration) patterns of characters to ignore? For example, we have a lot of posts that mention commit SHAs; we know those are "misspelled" but also cannot practically add all of them to the dictionary. So it would be nice if the dictionary supported patterns that could be ignored (let me know if I should create an issue).
1. Similarly, is there a way we could programatically generate dictionary entries? For example, all of our rule names show up in the dictionary and again it would be better if we could just dynamically add all of our rule filenames to the dictionary somehow.
1. Also, would it be possible to do spellcheck on individual words in hyphenates, again relating to our rule names? (e.g., don't check `space-infix-ops`, but instead check `space`, `infix`, and `ops` individually).

That said, this looks neat and interesting and I'd like to keep exploring this.

**Comment by tbroadley** [comment]
@platinumazure Thanks for the feedback! I really appreciate the time you've taken to look into this.

1. I think it would be simple to treat dictionary entries as regexes wrapped with `^` and `$` and whitelist words matched by the regexes. Then `[0-9a-f]{7}` could be added to the dictionary to whitelist short Git SHAs. I've opened https://github.com/tbroadley/spellchecker-cli/issues/13 to track this.
2. What about allowing multiple dictionaries to be specified, as a first step, then accepting JavaScript files as dictionaries? A JS dictionary would export an array of strings or RegExps, or maybe a function that returned such an array. For the use case of whitelisting rule names, perhaps the dictionary could instantiate a `Rules` object as defined in [`lib/rules.js`](https://github.com/eslint/eslint/blob/master/lib/rules.js) and export `rules.getAllLoadedRules().keys()`.
3. [`retext-spell`](https://github.com/retextjs/retext-spell), the library used by the tool to check spelling, doesn't support checking each part of a hyphenated word. It's possible to fork the library and change that. I suppose the benefit of checking each part would be to ensure that the names of newly-added rules are spelled correctly?

**Comment by platinumazure** [comment]
@tbroadley Multiple dictionaries could conceivably work although we definitely would want to be able to generate dictionary entries.

Re: hyphenation: Yes, you have correctly understood my intended purpose.... I wonder if it might be possible to create a new retext plugin which ignores all words except hyphenated words, and then sends the components of each hyphenate into retext-spell? We would still need to maintain our rule name dictionary (to avoid errors with the full rule names in retext-spell), but that is honestly mitigated with dynamic dictionary entries.

**Comment by tbroadley** [comment]
@platinumazure You're right, a retext plugin is probably the way to go. I'll investigate how I might implement that.

**Comment by platinumazure** [comment]
Hi @tbroadley, I've gone ahead and created a [POC branch](https://github.com/eslint/eslint/tree/spellcheck) to show how a dictionary might currently look. (N.B. I'm aware I haven't configured spellchecker in build scripts-- mostly I'm just putting up the dictionary.)

I generated the dictionary with the following command:

```bash
./node_modules/.bin/spellchecker --generate-dictionary "**/*.md" -i "[0-9a-f]{7,8}"
```

And it results in a dictionary about 1600 lines long-- a bit unwieldy, although a lot better than without the ignore pattern.

Some observations:

* The ignore patterns don't get into the dictionary (when combining `--generate-dictionary` and `-i`), wondering if they should?
* Right now, putting aside the length of the dictionary, there are very many classes of words in the dictionary. I mentioned rule names in the past as something I'd want to auto-generate some other way; another example I'm seeing now is contributor names.

I'll analyze the dictionary file more later and see if there are any other useful observations I could make.

**Comment by platinumazure** [comment]
@tbroadley Thanks for your prompt reply!

> It might make sense to change this behaviour to treat each line of the dictionary as a regular expression. That way, ignore patterns could be included in the dictionary instead of being passed on the command line. What do you think about that?

Assuming that doesn't absolutely murder performance, that sounds awesome to me.

> I added support for programmatic dictionaries to the [4.0 branch](https://github.com/tbroadley/spellchecker-cli/tree/4.0) of `spellchecker-cli`. Using that branch, it's possible to specify a JavaScript file that exports a list of strings to ignore as a dictionary.

Thanks for letting me know about that branch, I wasn't aware of it. I'll check it out when I get time!

**Comment by platinumazure** [comment]
I've been exploring this on and off, meaning I basically have been taking a crack at creating a local branch once in a while. Would it be possible to leave open another week or so and I can try this once more and provide feedback? If not, no worries about closing this. Thanks!

**Comment by not-an-aardvark** [comment]
Sure, fine by me -- just wanted to double-check that this is still being worked on.

**Comment by tbroadley** [comment]
@platinumazure FYI, the changes that were on the 4.0 branch have been merged into master.

**Comment by nzakas** [comment]
@platinumazure are you still pursuing this? If not, I think it's time to close this issue.

**Comment by platinumazure** [comment]
We can close this. I'm still interested but don't have time to keep digging for the short term.
